<launch>
    <arg name="rtabmap_workingDirectory" default="~/rtabmap_working_dir"/>
    <arg name="database_path"            default="~/rtabmap_working_dir/rtabmap.db"/>
    <arg name="initial_pose"      default="0 0 0 0 0 0"/>

    <arg name="subscribe_scan"   default="true"/>
    <arg name="scan_topic"       default="/scan"/>
    
    <arg name="cam_namespace"    default="/zedm/zed_node"/>
    <arg name="cfg"              default="$(find rtabmap_ros)/launch/config/rtabmap_config_zed_2dLidar.ini" /> <!--RTABMap parameters configuration-->


    <!-- sim time for convenience, if playing a rosbag set true -->
    <arg name="use_sim_time"            default="false"/>

    <!-- ********** RTABMAP LAUNCH ********** -->
    <include file="$(find rtabmap_ros)/launch/rtabmap.launch">
        <!-- <arg name="output"              value="log"/> -->
        <arg name="cfg"                     value="$(arg cfg)"/>
        <arg name="rtabmap_args"             value="--delete_db_on_start"/>
        <arg name="use_sim_time"             value="$(arg use_sim_time)"/>

        <arg name="visual_odometry"          value="false"/>
        <arg name="odom_topic"               value="$(arg cam_namespace)/odom"/>   <!-- Odometry topic name -->  
        <arg name="frame_id"                 value="base_link"/>  <!-- fixed body base_link --> 
        <arg name="initial_pose"             value="$(arg initial_pose)"/>
        <arg name="data_odom_sync"          value="true" />
        <arg name="subscribeOdomInfo"       value="true"/>

        <!-- Lidar related parameters -->
        <arg name="subscribe_scan"           value="$(arg subscribe_scan)"/>
        <arg name="scan_topic"               value="$(arg scan_topic)"/>

        <!-- RGB-D related parameters -->
        <arg name="rgb_topic"                value="$(arg cam_namespace)/rgb/image_rect_color"/>
        <arg name="camera_info_topic"        value="$(arg cam_namespace)/rgb/camera_info"/>
        <arg name="depth_topic"              value="$(arg cam_namespace)/depth/depth_registered"/>
        <arg name="depth_camera_info_topic"  value="$(arg cam_namespace)/depth/camera_info"/>
        <arg name="queue_size_sync"         value="30" />

        <!--mapping related parameters -->
        <arg name="rtabmap_workingDirectory" value="$(arg rtabmap_workingDirectory)"/>
        <arg name="database_path"            value="$(arg database_path)"/>
        <arg name="publish_tf_odom"         default="false"/>
        <arg name="publish_tf_map"          default="true"/>
        <arg name="raytracing"              default="true"/>
        <arg name="time_threshold"          default="700"/>  <!-- If the map takes longer than this threshold in milliseconds to update then it is moved to long term storage (does not impact octomap) -->
        <arg name="memory_threshold"        default="1000"/> <!--  "Rtabmap/TimeThr" : number of nodes in octomap before moved to long-term storage-->
        <arg name="map_resolution"          default="0.05"/>  <!-- "Grid/CellSize": resolution in m of grid/octomap-->
        <arg name="max_nodes"               default="1500"/> <!-- "GridGlobal/MaxNodes:" : maximum number of nodes in octomap-->
        <arg name="queue_size"              default="30"/>
        <arg name="rtabmap_detectionRate"   default="1"/>   <!-- "Rtabmap/DetectionRate" : [Detection rate (Hz). RTAB-Map will filter input images to satisfy this rate.] -->
        <arg name="tf_delay"                default="0.033" />  <!-- (~30 Hz) in order to keep up with the odometry time update-->
    </include>

</launch>